File: Python/event_system/event_system.py
from collections import defaultdict


class EventSystem:
    class Callback:
        def __init__(self, callback, order):
            self.callback = callback
            self.order = order

    def __init__(self):
        self.__callbacks = defaultdict(list)

    def subscribe(self, event, callback, order=0):
        callbacks: list = self.__callbacks[event]
        callbacks.append(EventSystem.Callback(callback, order))
        callbacks.sort(key=lambda c: c.order)

    def unsubscribe(self, event, callback):
        callbacks: list = self.__callbacks[event]
        callbacks[:] = [cb for cb in callbacks if cb.callback != callback]

    def publish(self, event):
        callbacks = self.__callbacks[type(event)]
        for callback in callbacks:
            callback.callback(event)

File: CSharp/EventSystem/EventSystem.cs
using System;
using System.Collections.Generic;

namespace EventSystem
{
    class EventSystem
    {
        class Callback
        {
            public MulticastDelegate Delegate { get; set; }
            public int Order { get; set; }
        }

        private Dictionary<Type, List<Callback>> _callbacks = new Dictionary<Type, List<Callback>>();

        public void Subscribe<TEvent>(Action<TEvent> callback, int order = 0) where TEvent : Event
        {
            if (_callbacks.TryGetValue(typeof(TEvent), out List<Callback> callbacks) == false)
            {
                callbacks = new List<Callback>();
                _callbacks.Add(typeof(TEvent), callbacks);
            }
            callbacks.Add(new Callback()
            {
                Delegate = callback,
                Order = order
            });
            callbacks.Sort((x, y) => x.Order - y.Order);
        }

        public void Publish<TEvent>(TEvent e) where TEvent : Event
        {
            if (_callbacks.TryGetValue(typeof(TEvent), out List<Callback> callbacks))
            {
                callbacks.ForEach(callback => ((Action<TEvent>)callback.Delegate)(e));
            }
        }
    }
}